### PR TITLE
[IMP] website: add warning to html/css editor

### PR DESCRIPTION
**Pull request by sleepy-monax**
Description of the feature this PR addresses:

This PR added a warning that overlay the code editor and tries to redirect them to the head/body editor popup.

Current behavior before PR:

Right the warning is a simple toast notification on top of the editor without actions for the user.

Desired behavior after PR is merged:

A an overlay with three button show up when the html editor is open, asking the if they really want to proceed.

2023-03-30-102732_721x327_scrot

task-3229205

### CODE REVIEW

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        callback: (ctx) => {
+            websiteId = ctx['website_id'];
+        },
+    });
+
+    let website = (await widget._rpc({
+        model: 'website',
+        method: 'read',
+        args: [[websiteId], ['custom_code_head', 'custom_code_footer']],
+    }))[0]
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:61]: There is a missing `;` here -> you have to install a linter. We have a tutorial to do that: https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines#tooling although I know there may have been problem with it in the past. Try to install using that and we'll see :+1:

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
+/** @odoo-module **/
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:1]: Commit message convention:
- Some info here: https://www.odoo.com/documentation/16.0/contributing/development/git_guidelines.html#commit-message-header
- Max 72 characters per line (title and message) (some editors actually set up a 72 chars ruler automatically, you could set one)
- No point at the end of the commit title
- The title must complete the sentence "when applied, this commit will ..."
- You commit should mention all the apps you touched (so not only web_editor here)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Small review before looking deeper :+1:

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import Dialog from 'web.Dialog';
+import core, { _t } from 'web.core';
+
+/**
+ * Render an code editor in the given node.
+ * @param {DOMElement} node where the editor will be rendered
+ * @param {String | undefined} content text of the editor
+ * @returns {Object} the ace editor instance
+ */
+export function renderCodeEditor(node, content) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:13]: Nice that you are adding a doc :+1: 
- Add a blank line between the description and the first `@param`
- DOMElement is not a type
- Instead of `{String | undefined} content`, write `{string} [content]` (optional param and built-in type params should be lowercase)

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    let website = (await widget._rpc({
+        model: 'website',
+        method: 'read',
+        args: [[websiteId], ['custom_code_head', 'custom_code_footer']],
+    }))[0]
+
+    const $content = $(core.qweb.render('web_editor.custom_code_dialog_content', {
+        headContentText: _t('Enter code that will be added into the <head> of every page of your site.'),
+        bodyContentText: _t('Enter code that will be added before the </body> of every page of your site.'),
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:65]: Why are those not part of the qweb template?

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    });
+
+    let websiteId;
+    widget.trigger_up('context_get', {
+        callback: (ctx) => {
+            websiteId = ctx['website_id'];
+        },
+    });
+
+    let website = (await widget._rpc({
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:57]: Instead of an util which receives the parent widget and then calls private method like this here, could not you make a specialized dialog instance which is in charge of doing this RPC when it is being opened? It's fine like this but if we can avoid calling private methods (the ones that start with _), it's better :+1: 

So instead of `openHeadBodyEditor(this)`, I would imagine something like `new HeadBodyDialog(this).open()` and all those libs loading, website read etc would be part of the Dialog lifecycle (willStart, start, open, ...).

File: addons/web_editor/static/src/js/common/ace-dialog.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        ],
+    });
+
+    await new Promise(resolve => {
+        dialog.on('closed', widget, resolve);
+        dialog.open();
+    });
+}
+
+
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace-dialog.js:109]: Only one new line at the end of files :+1:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -2,6 +2,8 @@ odoo.define('website.editor.snippets.options', function (require) {
 'use strict';
 
 const {ColorpickerWidget} = require('web.Colorpicker');
+
+var { openHeadBodyEditor } = require('@web_editor/js/common/ace-dialog');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:6]: In new code, we never use `var` anymore. Either `const` when you can (most of the time), or `let`.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -1260,6 +1262,7 @@ options.registry.OptionsTab = options.Class.extend({
         const widget = this._requestUserValueWidgets(params.imagepicker)[0];
         widget.enable();
     },
+
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:1265]: At least in legacy code (maybe it will be like that in OWL classes), don't add those extra blank lines between methods.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -1268,81 +1271,15 @@ options.registry.OptionsTab = options.Class.extend({
         await this.customizeWebsiteVariable(previewMode, this.bodyImageType, {variable: 'body-image-type'});
         await this.customizeWebsiteVariable(previewMode, widgetValue ? `'${widgetValue}'` : '', {variable: 'body-image'});
     },
+
     /**
+     * Directly hook OptionTab::openCustomCodeDialog to website.editor.custom_code::openCustomCodeDialog
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:1276]: I am not sure what this comment means, seems outdated

File: addons/website/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -1268,81 +1271,15 @@ options.registry.OptionsTab = options.Class.extend({
...
-            title = _t('Custom head code');
-            contentText = _t('Enter code that will be added into the <head> of every page of your site.');
-        } else {
-            fieldName = 'custom_code_footer';
-            title = _t('Custom end of body code');
-            contentText = _t('Enter code that will be added before the </body> of every page of your site.');
-        }
-
-        await Promise.all([libsProm, dataProm]);
+        return await openHeadBodyEditor(this)
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:1280]: `return await` almost never makes sense, you can just `return` I think here :+1:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -1455,35 +1392,7 @@ options.registry.OptionsTab = options.Class.extend({
...
-        });
-
-        const aceSession = aceEditor.getSession();
-        aceSession.setOptions({
-            mode: "ace/mode/qweb",
-            useWorker: false,
-        });
-        return aceEditor;
-    },
+    }
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:1395]: Detail but add a trailing `,`

File: addons/web_editor/static/src/xml/ace.xml [hunk 12]
```diff
@@ -49,13 +49,34 @@
...
-            <div class="alert alert-warning alert-dismissible mt-2 mb-0" role="alert">
-                Editing a built-in file through this editor is not advised, as it will prevent it from being updated during future App upgrades.
-                <button type="button" class="btn-close" aria-label="Close"></button>
+        </div>
+        <div id="ace-warning-message" class="mt-2 mb-0" role="alert">
+            <div class="inner">
+                <i class="fa fa-warning"></i>
+                <h1>HTML edits might be lost during future Odoo upgrade.</h1>
+                <p>If you need to add analytics or marketing tags, you can use code injection which is upgrade proof.</p>
+                <i>Note: To embed code in your page, use building block "Custom Code"</i>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/ace.xml:58]: In Odoo, we try to use `<i>` only for icons. Use `<em>` or bootstrap classes

File: addons/web_editor/static/src/xml/ace.xml [hunk 13]
```diff
@@ -49,13 +49,34 @@
...
         <div id="ace-view-editor"/>
     </div>
 </t>
 
+<!-- Theme - custom code -->
+<div t-name="web_editor.custom_code_dialog_content">
+    <div class="mb-2" t-esc="headContentText"/>
+    <div class="o_ace_head_editor_container"/>
+    <div class="mb-2" t-esc="bodyContentText"/>
+    <div class="o_ace_body_editor_container"/>
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/ace.xml:78]: Could feel overkill but since those two lines are `website.custom_code_dialog_content`, maybe you could move that template in web_editor and t-call it twice here? :+1:

File: addons/web_editor/static/src/js/common/ace.js [hunk 14]
```diff
@@ -229,6 +229,7 @@ var ViewEditor = Widget.extend({
      * @override
      */
     start: function () {
+        console.log("STARTING");
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace.js:232]: Oops?

File: addons/web_editor/static/src/js/common/ace.js [hunk 15]
```diff
@@ -242,7 +243,8 @@ var ViewEditor = Widget.extend({
         this.$includeBundlesArea = this.$('.oe_include_bundles');
         this.$includeAllSCSSArea = this.$('.o_include_all_scss');
         this.$viewID = this.$('#ace-view-id > span');
-        this.$warningMessage = this.$('#ace-view-id > .alert-warning');
+        this.$warningMessage = this.$('#ace-warning-message');
+        console.log("WARNING MESSAGE", this.$warningMessage);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace.js:247]: Oops?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Looks good but some mistakes to solve and things to fine-tune to make it perfect :+1:

File: addons/web_editor/static/src/js/common/ace-dialog.js (hunk 5 above)
Comment (sleepy-monax) [reply addons/web_editor/static/src/js/common/ace-dialog.js:57]: Great idea I'll look into it :)

File: addons/web_editor/static/src/js/common/ace-dialog.js (hunk 4 above)
Comment (sleepy-monax) [reply addons/web_editor/static/src/js/common/ace-dialog.js:65]: It wasn't in the old code, but yeah I should move them there

File: addons/web_editor/static/src/js/common/ace.js [hunk 16]
```diff
@@ -201,8 +201,8 @@ var ViewEditor = Widget.extend({
             defaultBundlesRestriction: [],
         });
 
-        this.resources = {xml: {}, scss: {}, js: {}};
-        this.editingSessions = {xml: {}, scss: {}, js: {}};
+        this.resources = { xml: {}, scss: {}, js: {} };
+        this.editingSessions = { xml: {}, scss: {}, js: {} };
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace.js:205]: Don't add such noise in this commit. One of two fixing liting is ok but I see many:
- It should be in its own commit
- Here I am not even sure we enforce that convention? Or is the Odoo linter configuration doing that?

File: addons/web_editor/static/src/js/common/ace.js [hunk 17]
```diff
@@ -946,7 +946,7 @@ var ViewEditor = Widget.extend({
     /**
      * Called when the close button is clicked -> hides the ace editor.
      *
-     * @private
+     * @privatedata
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace.js:949]: :thinking: ?

File: addons/website/static/src/services/website_service.js [hunk 18]
```diff
@@ -35,7 +35,7 @@ export const websiteService = {
         let lastUrl;
         let websiteRootInstance;
         let Wysiwyg;
-        let isRestrictedEditor;
+        let isRestrictedEditor;websiteService
```
Comment (qsm-odoo) [inline addons/website/static/src/services/website_service.js:38]: :thinking: Don't forget to review your PR yourself before marking it ready :wink:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Some new changes to make :)

File: addons/web_editor/__manifest__.py [hunk 19]
```diff
@@ -134,6 +135,8 @@
             'web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js',
             'web_editor/static/src/js/wysiwyg/fonts.js',
             'web_editor/static/src/xml/ace.xml',
+            'web_editor/static/src/xml/ace-dialog.xml',
```
[the remark this answers is not in the document]
Comment (sleepy-monax) [reply addons/web_editor/__manifest__.py]: `ace.xml` was blindly copied from `assets_common` to `assets_frontend` to eliminate `assets_common` in the long term. I didn't find any other usage of this template anywhere else, so moving it to assets_backend and removing it from `assets_frontend` and `assets_common` is safe.

File: addons/web_editor/static/src/js/common/ace.js (hunk 16 above)
Comment (rdeodoo) [reply addons/web_editor/static/src/js/common/ace.js]: > Or is the Odoo linter configuration doing that?

Is it @sleepy-monax ? I was going to suggest something in the Odoo task and saw this one, was interested by the answer as I am still using my own linter config.

**Comment by rdeodoo** [comment]
What's the status of this one?
I saw that you push forced after QSM review but no ping afterward and task is not in RD Dev Green.

**Comment by sleepy-monax** [comment]
@rdeodoo It's ready for review, I assumed @qsm-odoo was reviewing it since he set the state of the task to `Rd Review In progress` in odoo

**Comment by rdeodoo** [comment]
Indeed, I overlooked the status and thought it was "In review - Red" but it's in progress.

File: addons/web_editor/static/src/js/common/ace.js [hunk 20]
```diff
@@ -1112,6 +1117,11 @@ var ViewEditor = Widget.extend({
     _onEditAnywayClick: function () {
         this.$warningMessage.addClass('d-none');
     },
+
+    _onDontAskAgainClick: function () {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/ace.js:1121]: This last commit should be squashed with the main work, probably does not deserve a commit of its own :)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I planned to merge so I fine-tuned a few things and [push-forced](https://github.com/odoo/odoo/compare/592b0c15cc5591e67cf15a8fe9c18618b410fe5f..67286f1e61724716d8b0bbfa1e6c1473a2d74c06) so don't forget to fetch first :+1: 

In the end, there are a few more things to change to make this possible to merge :+1:

File: addons/web_editor/static/src/js/common/ace.js [hunk 21]
```diff
@@ -1042,10 +1042,74 @@ var ViewEditor = Widget.extend({
...
+            {
+                title: _t("Custom head/body code"),
+            },
+            [
+                {
+                    key: "custom_code_head",
+                    label: _t(
+                        "Enter code that will be added into the <head> of every page of your site."
+                    ),
+                    code: website["custom_code_head"] || "",
```
[the remark this answers is not in the document]
Comment (sleepy-monax) [reply addons/web_editor/static/src/js/common/ace.js]: I moved all the code related to this popup and the warning to website and rewrote the dialog in Owl as advised by @detrouxdev

File: addons/website/static/src/components/ace_warning/ace_warning.js [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+        this.dialog = useService("dialog");
+
+        this.state = useState({
+            visible: (localStorage.getItem("website.ace.doNotShowWarning") || "false") === "false",
+        });
+
+        this.closeEditor = this.closeEditor.bind(this);
+        this.hideWarning = this.hideWarning.bind(this);
+        this.dontAskAgain = this.dontAskAgain.bind(this);
+        this.injectCode = this.injectCode.bind(this);
```
Comment (qsm-odoo) [inline addons/website/static/src/components/ace_warning/ace_warning.js:21]: You seem to be the only code in Odoo to do that to work with `t-on-click` so my guess is that this is not necessary. There also seem to be quite a followed convention to make handler function start with `on` (just like old Widget where the convention was `_on`).

File: addons/website/static/src/components/ace_warning/ace_warning.js [hunk 23]
```diff
@@ -0,0 +1,41 @@
+/** @odoo-module **/
+
+import { EditHeadBodyDialog } from "../edit_head_body_dialog/edit_head_body_dialog";
+import { Component, useState } from "@odoo/owl";
+import { useService } from "@web/core/utils/hooks";
+
+export class AceEditorWarningOverlay extends Component {
+    static template = "website.AceEditorWarningOverlay";
+
+    setup() {
```
Comment (qsm-odoo) [inline addons/website/static/src/components/ace_warning/ace_warning.js:10]: Add docstring for all your methods

File: addons/website/static/src/components/ace_warning/ace_warning.xml [hunk 24]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+    <t t-name="website.AceEditorWarningOverlay" owl="1">
+        <div t-if="state.visible" class="position-absolute top-0 start-0 bottom-0 end-0 d-flex align-items-center justify-content-center z-3 p-4 bg-light bg-opacity-75" role="alert">
```
Comment (qsm-odoo) [inline addons/website/static/src/components/ace_warning/ace_warning.xml:4]: Don't indent the first level of XML files, add blank lines around

File: addons/website/static/src/components/ace_warning/ace_warning.xml [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+            </div>
+
+            <button type="button"
+                class="btn btn-link position-absolute bottom-0 end-0 m-4"
+                t-on-click="dontAskAgain">
+                Don't show again
+            </button>
+        </div>
+    </t>
+</templates>
```
Comment (qsm-odoo) [inline addons/website/static/src/components/ace_warning/ace_warning.xml:28]: Missing new line at the end of this file

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.js [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+import { useService } from "@web/core/utils/hooks";
+import { Component, onWillStart, useState } from "@odoo/owl";
+import { Dialog } from "@web/core/dialog/dialog";
+import { CodeEditor } from "@web/core/code_editor/code_editor";
+
+/**
+ * A dialog that let the user edit the code that will be injected
+ * in the <head> and before the </body> of every page of the website.
+ * This is a stable and upgrade proof alternative to directly editing
+ * the website xml
```
Comment (qsm-odoo) [inline addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.js:12]: Nitpicking: "." at the end of this line.

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.js [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+        this.website = useService("website");
+
+        this.state = useState({
+            head: "",
+            body: "",
+        });
+
+        this.onSave = this.onSave.bind(this);
+        onWillStart(async () => {
+            const ws = (
```
Comment (qsm-odoo) [inline addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.js:29]: Avoid small variable names like that, `website` would be clearer :+1:

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml [hunk 28]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+<t t-name="website.EditHeadBodyDialog" owl="1">
```
Comment (qsm-odoo) [inline addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml:3]: Blank lines around the t-name definiton

File: addons/website/static/src/js/editor/snippets.options.js [hunk 29]
```diff
@@ -1331,83 +1329,11 @@ options.registry.OptionsTab = options.Class.extend({
...
-                    {
-                        text: _t("Discard"),
-                        close: true,
-                    },
-                ],
-            });
-            dialog.on('closed', this, resolve);
-            dialog.open();
-        });
+        this.trigger_up('open_edit_head_body_dialog');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:1336]: Method options should only resolve their promise when the related action is finished. Here you trigger_up... then it is immediately synchronously finished. This is not ok because it means that other option can have their effect triggered as this one is considered finished. In general, this also means that your CTRL-Z entry for this option won't work (in this case as it is a setting in DB, it probably does not matter but...).

The common way to handle this with a trigger_up would be to give a function to call when the called action is done and have that function resolve a local promise defined here. Something like:
```javascript
return new Promise(resolve => {
    this.trigger_up('open_edit_head_body_dialog', {
        onSuccess: resolve,
    });
});
```

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

**Review by qsm-odoo — APPROVED** [review]
Rebased, fixed the many conflicts (owl changes, etc), cleaned and fixed some things, should be good to go now.

@robodoo rebase-ff r+

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml [hunk 30]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+<t t-name="website.EditHeadBodyDialog">
+    <Dialog title="'Edit Head and Body Code'">
+        <div>Enter code that will be added into the &lt;head&gt; of every page of your site.</div>
+        <CodeEditor mode="xml" theme="'monokai'"
```
Comment (aab-odoo) [inline addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml:7]: I think this should be `mode="'xml'"` not `mode="xml"`

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+            class="'my-2 rounded border o_website_code_editor_field'"
+            value="state.head"
+            onChange='v => state.head = v'/>
+
+        <div class="mt-2">Enter code that will be added before the &lt;/body&gt; of every page of your site.</div>
+        <CodeEditor mode="xml" theme="'monokai'"
```
Comment (aab-odoo) [inline addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml:13]: same here

File: addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml (hunk 30 above)
Comment (rdeodoo) [reply addons/website/static/src/components/edit_head_body_dialog/edit_head_body_dialog.xml:7]: KI guess it was noticed with https://github.com/odoo/odoo/pull/148063/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#diff-1e62b5d92d202f35c4233e53cdaf45af9aad79562fed319ec511666fb0f3ad02R13?

Could you backport it?
